src/odemis/acq/align/test/roi_autofocus_test.py

```diff
@@ -0,0 +1,136 @@
+"""
+import logging
+import os
+import time
+import unittest
+from concurrent import futures
+
+import odemis
+from odemis import model
+from odemis.acq.align.autofocus import estimateAutoFocusTime
+from odemis.acq.align.roi_autofocus import autofocus_in_roi, estimate_autofocus_in_roi_time
+from odemis.util import testing
+
+logging.getLogger().setLevel(logging.DEBUG)
+logging.basicConfig(format="%(asctime)s  %(levelname)-7s %(module)s:%(lineno)d %(message)s")
+
+CONFIG_PATH = os.path.dirname(odemis.__file__) + "/../../install/linux/usr/share/odemis/"
+METEOR_CONFIG = CONFIG_PATH + "sim/meteor-sim.odm.yaml"
+
+
+class RoiAutofocusTestCase(unittest.TestCase):
+    """
+    Unit test for the roi_autofocus module
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        testing.start_backend(METEOR_CONFIG)
+
+        # create some streams connected to the backend
+        # cls.microscope = model.getMicroscope()
+        cls.ccd = model.getComponent(role="ccd")
+        cls.focus = model.getComponent(role="focus")
+        cls.stage = model.getComponent(role="stage")
+
+        # Assumes the stage is referenced
+        cls.init_pos = (cls.stage.position.value["x"], cls.stage.position.value["y"])
+        cls.width, cls.height = cls.ccd.shape[:2]
+        cls.px_size = cls.ccd.getMetadata().get(model.MD_PIXEL_SIZE)
+
+        # Assume initial focus position is good
+        # The initial position comes from manual focussing the centre of overview image
+        current_focus = cls.focus.position.value["z"]  # assume manual focus before running the script
+        # Allow a range of +/- 30 µm around the focused position
+        cls.focus_range = (current_focus - 30.0e-6, current_focus + 30.0e-6)
+
+    @classmethod
+    def tearDownClass(cls):
+        logging.debug("moving back to initial position")
+        cls.stage.moveAbs({"x": cls.init_pos[0], "y": cls.init_pos[1]}).result()
+
+    def test_autofocus_in_roi(self):
+        """
+        Tests the autofus in roi is running and returns the focus points:
+        """
+        init_pos = self.init_pos
+        px_size = self.px_size
+        width = self.width
+        height = self.height
+        n_tiles = (3, 3)
+        overlap = 0.2
+
+        xmin = init_pos[0] - (1 - overlap) * n_tiles[0] / 2 * px_size[0] * width
+        ymin = init_pos[1] - (1 - overlap) * n_tiles[1] / 2 * px_size[1] * height
+        xmax = init_pos[0] + (1 - overlap) * n_tiles[0] / 2 * px_size[0] * width
+        ymax = init_pos[1] + (1 - overlap) * n_tiles[1] / 2 * px_size[1] * height
+
+        bbox = (xmin, ymin, xmax, ymax)
+
+        f = autofocus_in_roi(bbox, self.stage, self.ccd, self.focus, self.focus_range, n_tiles[0], n_tiles[1], 0)
+
+        # Test if the autofocus in roi is running
+        time.sleep(0.1)
+        self.assertTrue(f.running())
+
+        # The returned focus points should be minimum three
+        focus_points = f.result()
+        self.assertGreaterEqual(len(focus_points), 3)
```

Comment:
What should be the format of each focus point? (x,y, z)? Please also check that. 

BTW, is there a maximum amount of focus points that should be measured? (eg 9), if so, you could check that too.

Reply:
At the moment we have not defined maximum number of focus points required for autofocus. The 9 points we take initially can be reduced to a lower number if the confidence level when the focus point is below a confidence level. (conf_level = 0.8). More than 9 focus points can be included in future depending on the number of tiles. In both the cases, minimum number of focus points can be checked.  

I have added the following
`        # The returned focus points should be minimum three
        # One focus point has three coordinates x,y and z
        # The returned focus points should be minimum three to fit a plane for the autofocus`